File: esphome/components/alpha3/select.py
import esphome.codegen as cg
from esphome.components import select
import esphome.config_validation as cv
from esphome.const import ENTITY_CATEGORY_CONFIG

from . import Alpha3, alpha3_ns

CONF_ALPHA3_ID = "alpha3_id"

Alpha3Select = alpha3_ns.class_("Alpha3Select", select.Select, cg.Component)

CONF_PUMP_MODE = "pump_mode"

MODE_OPTIONS = [
    "AutoAdapt",
    "Constant Pressure 1 (Min)",
    "Constant Pressure 2",
    "Constant Pressure 3 (Max)",
    "Proportional Pressure 1 (Min)",
    "Proportional Pressure 2",
    "Proportional Pressure 3 (Max)",
    "Constant Frequency",
]

CONFIG_SCHEMA = cv.Schema(
    {
        cv.GenerateID(CONF_ALPHA3_ID): cv.use_id(Alpha3),
        cv.Optional(CONF_PUMP_MODE): select.select_schema(
            Alpha3Select,
            entity_category=ENTITY_CATEGORY_CONFIG,
        ),
    }
)


async def to_code(config):
    parent = await cg.get_variable(config[CONF_ALPHA3_ID])

    if pump_mode_config := config.get(CONF_PUMP_MODE):
        sel = await select.new_select(
            pump_mode_config,
            options=MODE_OPTIONS,
        )
        await cg.register_component(sel, pump_mode_config)
        cg.add(sel.set_parent(parent))

File: esphome/components/alpha3/select.cpp
#include "select.h"
#include "esphome/core/log.h"

#ifdef USE_ESP32

namespace esphome {
namespace alpha3 {

static const char *const TAG = "alpha3.select";

void Alpha3Select::control(const std::string &value) {
  if (this->parent_ == nullptr) {
    ESP_LOGE(TAG, "Parent not set");
    return;
  }

  ESP_LOGI(TAG, "Setting pump mode to: %s", value.c_str());

  if (value == "AutoAdapt") {
    this->parent_->set_mode_autoadapt();
  } else if (value == "Constant Pressure 1 (Min)") {
    this->parent_->set_mode_const_pressure(1);
  } else if (value == "Constant Pressure 2") {
    this->parent_->set_mode_const_pressure(2);
  } else if (value == "Constant Pressure 3 (Max)") {
    this->parent_->set_mode_const_pressure(3);
  } else if (value == "Proportional Pressure 1 (Min)") {
    this->parent_->set_mode_prop_pressure(1);
  } else if (value == "Proportional Pressure 2") {
    this->parent_->set_mode_prop_pressure(2);
  } else if (value == "Proportional Pressure 3 (Max)") {
    this->parent_->set_mode_prop_pressure(3);
  } else if (value == "Constant Frequency") {
    this->parent_->set_mode_const_freq();
  } else {
    ESP_LOGW(TAG, "Unknown mode: %s", value.c_str());
    return;
  }

  this->publish_state(value);
}

}  // namespace alpha3
}  // namespace esphome

#endif
